cmd/erasure-server-pool-decom.go

```diff
@@ -0,0 +1,881 @@
+
+	// Verbose information
+	ItemsDecommissioned     uint64 `json:"-" msg:"id"`
+	ItemsDecommissionFailed uint64 `json:"-" msg:"idf"`
+	BytesDone               uint64 `json:"-" msg:"bd"`
+	BytesFailed             uint64 `json:"-" msg:"bf"`
+}
+
+// bucketPop should be called when a bucket is done decommissioning.
+// Adds the bucket to the list of decommissioned buckets and updates resume numbers.
+func (pd *PoolDecommissionInfo) bucketPop(bucket string) {
+	pd.DecommissionedBuckets = append(pd.DecommissionedBuckets, bucket)
+	for i, b := range pd.QueuedBuckets {
+		if b == bucket {
+			// Bucket is done.
+			pd.QueuedBuckets = append(pd.QueuedBuckets[:i], pd.QueuedBuckets[i+1:]...)
+		}
+	}
+}
+
+func (pd *PoolDecommissionInfo) bucketsToDecommission() []string {
+	return pd.QueuedBuckets
+}
+
+func (pd *PoolDecommissionInfo) isBucketDecommissioned(bucket string) bool {
+	for _, b := range pd.DecommissionedBuckets {
+		if b == bucket {
+			return true
+		}
+	}
+	return false
+}
+
+func (pd *PoolDecommissionInfo) bucketPush(bucket string) {
+	var found bool
+	for _, b := range pd.QueuedBuckets {
+		if b == bucket {
+			found = true
+			break
+		}
+	}
+	if !found {
+		pd.QueuedBuckets = append(pd.QueuedBuckets, bucket)
+	}
+}
+
+// PoolStatus captures current pool status
+type PoolStatus struct {
+	ID           int                   `json:"id" msg:"id"`
+	CmdLine      string                `json:"cmdline" msg:"cl"`
+	LastUpdate   time.Time             `json:"lastUpdate" msg:"lu"`
+	Decommission *PoolDecommissionInfo `json:"decommissionInfo,omitempty" msg:"dec"`
+}
+
+//go:generate msgp -file $GOFILE -unexported
+type poolMeta struct {
+	Version int          `msg:"v"`
+	Pools   []PoolStatus `msg:"pls"`
+}
+
+func (p *poolMeta) DecommissionComplete(idx int) bool {
+	p.Pools[idx].LastUpdate = time.Now().UTC()
+	if p.Pools[idx].Decommission != nil {
+		p.Pools[idx].Decommission.Complete = true
+		p.Pools[idx].Decommission.Failed = false
+		p.Pools[idx].Decommission.Canceled = false
+		return true
+	}
+	return false
+}
+
+func (p *poolMeta) DecommissionFailed(idx int) bool {
+	p.Pools[idx].LastUpdate = time.Now().UTC()
+	if p.Pools[idx].Decommission != nil {
+		p.Pools[idx].Decommission.StartTime = time.Time{}
+		p.Pools[idx].Decommission.Complete = false
+		p.Pools[idx].Decommission.Failed = true
+		p.Pools[idx].Decommission.Canceled = false
+		return true
+	}
+	return false
+}
+
+func (p *poolMeta) DecommissionCancel(idx int) bool {
+	p.Pools[idx].LastUpdate = time.Now().UTC()
+	if p.Pools[idx].Decommission != nil {
+		p.Pools[idx].Decommission.StartTime = time.Time{}
+		p.Pools[idx].Decommission.Complete = false
+		p.Pools[idx].Decommission.Failed = false
+		p.Pools[idx].Decommission.Canceled = true
+		return true
+	}
+	return false
+}
+
+func (p poolMeta) IsBucketDecommissioned(idx int, bucket string) bool {
+	if p.Pools[idx].Decommission == nil {
+		// Decommission not in progress, return 'true' to
+		// avoid any decommissioning.
```

Comment:
This comment is not clear to me. Could you give an example situation where we may need to indicate that a given bucket is being decommissioned while `poolMeta` has no such information in it?

Reply:
When you are resuming a decommission from a failed or canceled operation we need to restart from where it was last failed.

Reply:
However this should return 'false' to decommission not the other way round.

Reply:
Actually no - so this code is meant for a situation where we check if bucket is decommissioned only when decommission is in progress.

Reply:
this code path should never be triggered otherwise.